controller/consoleapplication_controller.go

```diff
@@ -0,0 +1,131 @@
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *ConsoleApplicationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+	logger.Info("Reconciling ConsoleApplication...")
+
+	// Retrieving the ConsoleApplication CR instance requested for reconciliation
+	operatorCR := &appsv1alpha1.ConsoleApplication{}
+	if err := r.Get(ctx, req.NamespacedName, operatorCR); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Error(err, "ConsoleApplication Not Found")
+			return NoRequeue()
+		}
+		logger.Error(err, "Unable to fetch ConsoleApplication CR")
+		SetDegraded(operatorCR, appsv1alpha1.ReasonOperatorResourceNotAvailable.String(), err.Error())
+		if err := r.Status().Update(ctx, operatorCR); err != nil {
+			return RequeueWithError(err)
+		}
+		return RequeueOnError(err)
+	}
+
+	init := operatorCR.Status.Conditions == nil
+	if init {
+		operatorCR.Status.Conditions = make([]metav1.Condition, 0)
+		SetStarted(operatorCR)
+		if err := r.Status().Update(ctx, operatorCR); err != nil {
+			return RequeueWithError(err)
+		}
+	}
+
+	// Fetching the secret resource if specified in the CR
+	secretResourceName := operatorCR.Spec.Git.SourceSecretRef
+	decodedSecret := ""
+	if secretResourceName != "" {
+		secret := &corev1.Secret{}
+		if err := r.Get(ctx, client.ObjectKey{
+			Namespace: req.Namespace,
+			Name:      secretResourceName,
+		}, secret); err != nil {
+			logger.Error(err, "Unable To Find Secret Resource")
+			SetFailed(operatorCR, appsv1alpha1.ReasonSecretResourceNotFound.String(), err.Error())
+			if err := r.Status().Update(ctx, operatorCR); err != nil {
+				return RequeueWithError(err)
+			}
+			return NoRequeue()
+		}
+		// Decoding the secret data - Considering the Basic Auth secret
+		decodedSecret = string(secret.Data["password"])
+		logger.Info("Secret Decoded: " + decodedSecret)
+	}
+
+	// Checking if the Git Repository is reachable
+	gs := gitservice.New(operatorCR.Spec.Git.Url, operatorCR.Spec.Git.Reference, decodedSecret, logger)
+	g_status, g_reason := gs.IsRepoReachable()
+	logger.Info("Git Repository Reachable: " + string(g_status))
+
+	SetGitServiceCondition(operatorCR, g_status, g_reason.String())
+	if err := r.Status().Update(ctx, operatorCR); err != nil {
+		return RequeueWithError(err)
+	}
+
+	if g_status != metav1.ConditionTrue {
+		SetFailed(operatorCR, g_reason.String(), fmt.Sprintf("Git Repository Not Reachable: %s", g_reason.String()))
+		if err := r.Status().Update(ctx, operatorCR); err != nil {
+			return RequeueWithError(err)
+		}
+		return NoRequeue()
+	}
+
+	// Add the Strategy Service here: Return the list of resources config that needs to be created
+
+	logger.Info("All done!")
+	SetSucceeded(operatorCR)
+	if err := r.Status().Update(ctx, operatorCR); err != nil {
+		return RequeueWithError(err)
+	}
```

Comment:
I see we set the status 3 times here. Maybe we can avoid the status update at line 102 and just set it for failure reason or when all completed ?

Reply:
Here the idea was to inform the status of the GitRepo in the beginning itself while the controller is still creating the other resources like the Builds, Deployments, etc. This can help users take decisions faster.